rails_programming/hotwire/turbo.md

```diff
@@ -0,0 +1,587 @@
+selected content from our destination and inserted it into our current page!
+(But using our previous knowledge of Turbo Drive, we can advance our history using `data-turbo-action` if we desire.)
+
+#### Breaking out of a Turbo Frame
+
+Sometimes you may have a link inside of the Turbo Frame that you want to act as a normal page navigation. To do so, just add 
+`data-turbo-frame="_top"` to the element. An example with a Rails link helper:
+
+~~~erb
+<%= link_to "Return to Article", @article, data: { turbo_frame: "_top" } %>
+~~~
+
+#### Targeting a Turbo Frame from Outside
+
+We can also do the opposite! We can make a link that exists outside of our Turbo Frame act as if it was inside of the Frame and update it. This time, we set
+the `turbo-frame` data attribute to point to the ID of the specific frame. Lets say we want to designate a turbo frame to show either a list of posts or a list of images:
+
+~~~erb
+<%= link_to "Show Posts", posts_path, data: { turbo_frame: "list-region" } %>
+<%= link_to "Show Images", images_path, data: { turbo_frame: "list-region" } %>
+
+<%= turbo_frame id="list-region" %>
+~~~
+
+Clicking either of the above links will send a request to the respective path and return the content inside of our `"list-region"` frame.
+
+#### Src & Lazy Loading
+
+Frames can be given a `src` attribute. When this is supplied, the frame will be populated after the initial page load by making a separate request to the
+associated path. We can also wrap placeholder content inside of this `src` frame to create a [skeleton placeholder](https://uxdesign.cc/what-you-should-know-about-skeleton-screens-a820c45a571a).
+
+For example:
+
+~~~erb
+...
+<%= turbo_frame_tag id="Articles", src: articles_path do %>
+  <div>
+    I am a placeholder! After the request to articles_path is finished,
+    I will be replaced with the content inside of that page's turbo frame
+    with the id of "Articles".
+  </div>
+<% end %>
+~~~
+
+We can also make our frames **lazy loaded**. A lazy loaded frame will only fetch its content when it becomes visible on the page. We make a frame lazy load by adding
+`loading=lazy`. A frame with `loading=lazy` *must* also have a `src:` attribute to fetch from, or else it will not do anything. Using our previous example:
+
+~~~erb
+...
+<%= turbo_frame_tag id="Articles", src: articles_path, loading: "lazy" do %>
+  <div>
+    I am a placeholder! I will be replaced when a user scrolls down to see me on the page!
+  </div>
+<% end %>
+~~~
+
+### Turbo Stream
+
+Now we know how to set up our views to use Turbo Frames, but what about content that is being changed by our users? We can't put a Turbo Frame around
+something that doesn't exist yet! That's where Turbo Stream comes in. Turbo Streams send page changes as HTML wrapped in `<turbo-stream>` elements.
+Turbo streams have an *action* and a *target ID* to act the action onto. These streams can be sent in response to either a direct browser
```

Comment:
I'm not too sure why I phrased it in that way! Perhaps it would be better to phrase it more along the lines of 

> Turbo Streams specify an *action* to perform and the *target ID* of the DOM element to update with this action.

Reply:
Yeah that would read better IMO